File: sdk/golang/signal_rules.go
package prismer

import (
	"regexp"
	"strings"
)

// errorPattern maps a regex to a normalized error type.
type errorPattern struct {
	pattern *regexp.Regexp
	errType string
}

var errorPatterns = []errorPattern{
	{regexp.MustCompile(`(?i)timeout|timed?\s*out|deadline\s*exceeded|context\s*deadline`), "timeout"},
	{regexp.MustCompile(`(?i)econnrefused|connection\s*refused`), "connection_refused"},
	{regexp.MustCompile(`(?i)enotfound|dns|getaddrinfo|resolve`), "dns_error"},
	{regexp.MustCompile(`(?i)rate\s*limit|too\s*many\s*requests|429`), "rate_limit"},
	{regexp.MustCompile(`(?i)401|unauthorized|unauthenticated|auth.*fail`), "auth_error"},
	{regexp.MustCompile(`(?i)403|forbidden|access\s*denied|permission`), "permission_error"},
	{regexp.MustCompile(`(?i)404|not\s*found`), "not_found"},
	{regexp.MustCompile(`(?i)5\d{2}|internal\s*server|server\s*error|502|503|504`), "server_error"},
	{regexp.MustCompile(`(?i)type\s*error|typeerror`), "type_error"},
	{regexp.MustCompile(`(?i)syntax\s*error|syntaxerror|unexpected\s*token`), "syntax_error"},
	{regexp.MustCompile(`(?i)reference\s*error|referenceerror|is\s*not\s*defined`), "reference_error"},
	{regexp.MustCompile(`(?i)out\s*of\s*memory|oom|heap|allocation\s*failed`), "oom"},
	{regexp.MustCompile(`(?i)crash|panic|segfault|sigsegv|sigabrt`), "crash"},
	{regexp.MustCompile(`(?i)quota|limit\s*exceeded|insufficient`), "quota_exceeded"},
	{regexp.MustCompile(`(?i)tls|ssl|certificate|cert\s*verify`), "tls_error"},
	{regexp.MustCompile(`(?i)deadlock|lock\s*timeout|lock\s*wait`), "deadlock"},
}

// normalizeRegex is used to strip non-alphanumeric chars in the fallback path.
var normalizeRegex = regexp.MustCompile(`[^a-z0-9_]`)

// SignalExtractionContext holds the execution context for signal extraction.
type SignalExtractionContext struct {
	Error          string
	TaskStatus     string
	TaskCapability string
	Provider       string
	Stage          string
	Severity       string
	Tags           []string
}

// ExtractSignals extracts structured SignalTag list from execution context.
// Pure regex rules, zero external deps.
func ExtractSignals(ctx SignalExtractionContext) []SignalTag {
	var result []SignalTag

	// Error pattern matching
	if ctx.Error != "" {
		matched := false
		for _, ep := range errorPatterns {
			if ep.pattern.MatchString(ctx.Error) {
				tag := SignalTag{Type: "error:" + ep.errType}
				if ctx.Provider != "" {
					tag.Provider = ctx.Provider
				}
				if ctx.Stage != "" {
					tag.Stage = ctx.Stage
				}
				if ctx.Severity != "" {
					tag.Severity = ctx.Severity
				}
				result = append(result, tag)
				matched = true
				break
			}
		}
		if !matched {
			normalized := ctx.Error
			if len(normalized) > 50 {
				normalized = normalized[:50]
			}
			normalized = strings.ToLower(normalized)
			normalized = normalizeRegex.ReplaceAllString(normalized, "_")
			tag := SignalTag{Type: "error:" + normalized}
			if ctx.Provider != "" {
				tag.Provider = ctx.Provider
			}
			if ctx.Stage != "" {
				tag.Stage = ctx.Stage
			}
			result = append(result, tag)
		}
	}

	// Task status
	if ctx.TaskStatus == "failed" {
		result = append(result, SignalTag{Type: "task.failed"})
	} else if ctx.TaskStatus == "completed" {
		result = append(result, SignalTag{Type: "task.completed"})
	}

	// Capability
	if ctx.TaskCapability != "" {
		result = append(result, SignalTag{Type: "capability:" + ctx.TaskCapability})
	}

	// Custom tags
	for _, t := range ctx.Tags {
		result = append(result, SignalTag{Type: t})
	}

	return result
}

File: sdk/rust/src/signal_rules.rs
//! Signal Enrichment — SDK-side signal extraction using regex rules.
//!
//! Ported from server's signal-extractor.ts. 16 error patterns.
//! Zero external deps beyond std, synchronous.

use crate::types::SignalTag;

/// Execution context for signal extraction.
pub struct SignalExtractionContext {
    pub error: Option<String>,
    pub task_status: Option<String>,
    pub task_capability: Option<String>,
    pub provider: Option<String>,
    pub stage: Option<String>,
    pub severity: Option<String>,
    pub tags: Vec<String>,
}

impl Default for SignalExtractionContext {
    fn default() -> Self {
        Self {
            error: None,
            task_status: None,
            task_capability: None,
            provider: None,
            stage: None,
            severity: None,
            tags: Vec::new(),
        }
    }
}

struct ErrorPattern {
    keywords: &'static [&'static str],
    signal_type: &'static str,
}

const ERROR_PATTERNS: &[ErrorPattern] = &[
    ErrorPattern { keywords: &["timeout", "timed out", "timedout", "deadline exceeded", "context deadline"], signal_type: "timeout" },
    ErrorPattern { keywords: &["econnrefused", "connection refused"], signal_type: "connection_refused" },
    ErrorPattern { keywords: &["enotfound", "dns", "getaddrinfo", "resolve"], signal_type: "dns_error" },
    ErrorPattern { keywords: &["rate limit", "too many requests", "429"], signal_type: "rate_limit" },
    ErrorPattern { keywords: &["401", "unauthorized", "unauthenticated"], signal_type: "auth_error" },
    ErrorPattern { keywords: &["403", "forbidden", "access denied", "permission"], signal_type: "permission_error" },
    ErrorPattern { keywords: &["404", "not found"], signal_type: "not_found" },
    ErrorPattern { keywords: &["500", "502", "503", "504", "internal server", "server error"], signal_type: "server_error" },
    ErrorPattern { keywords: &["typeerror", "type error"], signal_type: "type_error" },
    ErrorPattern { keywords: &["syntaxerror", "syntax error", "unexpected token"], signal_type: "syntax_error" },
    ErrorPattern { keywords: &["referenceerror", "reference error", "is not defined"], signal_type: "reference_error" },
    ErrorPattern { keywords: &["out of memory", "oom", "heap", "allocation failed"], signal_type: "oom" },
    ErrorPattern { keywords: &["crash", "panic", "segfault", "sigsegv", "sigabrt"], signal_type: "crash" },
    ErrorPattern { keywords: &["quota", "limit exceeded", "insufficient"], signal_type: "quota_exceeded" },
    ErrorPattern { keywords: &["tls", "ssl", "certificate", "cert verify"], signal_type: "tls_error" },
    ErrorPattern { keywords: &["deadlock", "lock timeout", "lock wait"], signal_type: "deadlock" },
];

/// Extract structured SignalTag list from execution context.
/// Pure keyword matching, zero external deps, synchronous.
pub fn extract_signals(ctx: &SignalExtractionContext) -> Vec<SignalTag> {
    let mut result = Vec::new();

    if let Some(error) = &ctx.error {
        let lower = error.to_lowercase();
        let mut matched = false;

        for pattern in ERROR_PATTERNS {
            if pattern.keywords.iter().any(|kw| lower.contains(kw)) {
                let tag = SignalTag {
                    signal_type: format!("error:{}", pattern.signal_type),
                    provider: ctx.provider.clone(),
                    stage: ctx.stage.clone(),
                    severity: ctx.severity.clone(),
                };
                result.push(tag);
                matched = true;
                break;
            }
        }

        if !matched {
            let normalized: String = lower.chars().take(50)
                .map(|c| if c.is_alphanumeric() || c == '_' { c } else { '_' })
                .collect();
            result.push(SignalTag {
                signal_type: format!("error:{}", normalized),
                provider: ctx.provider.clone(),
                stage: ctx.stage.clone(),
                severity: None,
            });
        }
    }

    if let Some(status) = &ctx.task_status {
        match status.as_str() {
            "failed" => result.push(SignalTag { signal_type: "task.failed".to_string(), provider: None, stage: None, severity: None }),
            "completed" => result.push(SignalTag { signal_type: "task.completed".to_string(), provider: None, stage: None, severity: None }),
            _ => {}
        }
    }

    if let Some(cap) = &ctx.task_capability {
        result.push(SignalTag { signal_type: format!("capability:{}", cap), provider: None, stage: None, severity: None });
    }

    for tag in &ctx.tags {
        result.push(SignalTag { signal_type: tag.clone(), provider: None, stage: None, severity: None });
    }

    result
}
